### ISSUE TITLE
Preamble isn't compatible with Angluar 6+

### ISSUE THREAD

**Issue by lexaknyazev** [mbullington/node_preamble.dart#15]
Starting with version 6, Angular (TS, not Dart) deliberately disables shims for node-specific objects. This leads to packages using `node_preamble.dart` failing at runtime with:
```
Uncaught ReferenceError: global is not defined
```
on the first line of the preamble
```javascript
var self = Object.create(global);
```

Besides `global`, these variables are also undefined:
- `process`
- `__dirname`
- `__filename`

To fix this, we could
- change the first line to
  ```js
  var self = self || Object.create(global);
  ```
- and move
https://github.com/mbullington/node_preamble.dart/blob/7b3425ac52b745a294d1d14c1d64238f12a7fca5/lib/preamble.js#L15-L18
after https://github.com/mbullington/node_preamble.dart/blob/7b3425ac52b745a294d1d14c1d64238f12a7fca5/lib/preamble.js#L22

**Comment by lexaknyazev** [comment]
@mbullington any thoughts on this?

**Comment by mbullington** [comment]
Hi @lexaknyazev, sorry for the late response! I'd like a little more context for the problem:

How is Angular disabling this specifically? I don't have much experience with Angular but have used TypeScript quite a bit, how does the build system work? Would it be possible to exclude the projects using node_preamble from the tsconfig?

**Comment by lexaknyazev** [comment]
My project is written in Dart and an npm package is one of its deliverables. Some people may wish to use the npm package as a dependency in their Angular TS projects.

AFAIU, Angular uses webpack configured to disable shims for node-specific objects. So variables like `global` are not defined.

**Comment by mbullington** [comment]
I took a quick look at this, sorry for the lack of response & hold-up.

I'm still having trouble pinpointing down the issue, would I be able to download your package and make a simple Angular TS project with it? If you had an example that would be best. I feel like looking at the code it generates (since this is a runtime issue) would help.

In the mean time, I think a solution based on the node-detector you put in one of your previous tickets could be used, since some browser environments do polyfill process, __dirname, and __filename.

Thanks!

**Comment by mbullington** [comment]
Made an intended fix in another branch (https://github.com/mbullington/node_preamble.dart/tree/angular-6-fix), are you able to test the preamble like this?

**Comment by lexaknyazev** [comment]
Sorry for the delay. There are two issues with the updated preamble.

1. `global` is still used directly in two places: `scheduleImmediate` polyfill and `if (!global.window) {`. Replacing `global` with `self` in these two places helps.

2. `webpack` sometimes over-eagerly optimizes `typeof`, so that 
   ```js
   "undefined"!=typeof __dirname&&(self.__dirname=__dirname),"undefined"!=typeof __filename&&(self.__filename=__filename)
   ```
   becomes
   ```js
   true&&(self.__dirname=__dirname), true&&(self.__filename=__filename)
   ```
   and fails at runtime. A workaround found in sass.js source: https://github.com/medialize/sass.js/commit/6241c10dbe35b181ad78ecbd692fed5f86213ffb.

**Comment by mbullington** [comment]
I think I was able to fix these problems, please try the branch again and let me know if it works. Thanks!

**Comment by mbullington** [comment]
@lexaknyazev Hi, were you able to test these changes?

**Comment by lexaknyazev** [comment]
Apparently, this issue has many layers.

The main issue was with those variables that don't exist in the web environment. It is fixed now.

The secondary issue was with webpack breaking certain checks (that were enhanced with additional "guards"). This webpack issue has been finally fixed a few days ago, so as long as its version is `^4.32.2`, those "guards" aren't needed. This applies only to projects that use webpack directly.

Angular projects, however, use webpack indirectly, via `angular-cli` package that fixes a specific webpack version. The latest published `angular-cli` (`8.0.1`) is fixed on the webpack version `4.30.0` that is affected by the issue. Although, the `master` branch has already been upgraded to `4.32.2`.

With that said, I think it may be reasonable to remove those "guards" and document the minimum required webpack version. Angular projects using `angular-cli` would need to include this workaround in their `src/polyfills.ts` for some time:
```typescript
(window as any).__dirname = null;
(window as any).__filename = null;
```

**Comment by lexaknyazev** [comment]
@mbullington 
Stable version of Angular JS CLI is now fixed, so those separate vars (lines 7, 26, 31, and 36) are no longer needed. Could you please update the branch and merge changes?

**Comment by mbullington** [comment]
@lexaknyazev Will do, thanks for keeping updated on this ticket

**Comment by mbullington** [comment]
Should be included in v1.4.5, please let me know if this works for you. Thanks!
